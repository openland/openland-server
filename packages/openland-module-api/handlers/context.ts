import * as express from 'express';
import { Modules } from 'openland-modules/Modules';
import { createTracer } from 'openland-log/createTracer';
import { createLogger } from 'openland-log/createLogger';
import { createEmptyContext } from 'openland-utils/Context';
import { AuthContext } from 'openland-module-auth/AuthContext';
import { TracingContext } from 'openland-log/TracingContext';
import { CacheContext } from 'openland-module-api/CacheContext';
import { AppContext } from 'openland-modules/AppContext';

let tracer = createTracer('express');
const logger = createLogger('http');

async function context(src: express.Request): Promise<AppContext> {

    let res = createEmptyContext();
    let uid: number | undefined;
    let tid: string | undefined;
    let oid: number | undefined;

    // User
    if (src.user !== null && src.user !== undefined) {
        if (typeof src.user.sub === 'string') {
            uid = await Modules.Users.findUserByAuthId(createEmptyContext(), src.user.sub);
        } else if (typeof src.user.uid === 'number' && typeof src.user.tid === 'number') {
            uid = src.user.uid;
            tid = src.user.tid;
        }
    }
    // Organization
    if (uid) {
        let accounts = await Modules.Orgs.findUserOrganizations(createEmptyContext(), uid);

        // Default behaviour: pick the default one
        if (accounts.length >= 1) {
            oid = accounts[0];

            let profile = await Modules.Users.profileById(createEmptyContext(), uid);
            oid = (profile && profile.primaryOrganization) || oid;
        }
    }

    // Auth Context
    res = AuthContext.set(res, { tid, uid, oid });

    // Tracing Context
    res = TracingContext.set(res, { span: tracer.startSpan('http') });
    res = CacheContext.set(res, new Map());
    return new AppContext(res);
}

export async function callContextMiddleware(isTest: boolean, req: express.Request, res: express.Response, next: express.NextFunction) {
    let ctx: AppContext;
    try {
        ctx = await context(req);
    } catch (e) {
        res!!.status(404).send('Unable to find domain');
        return;
    }
    if (!isTest) {
        if (AuthContext.get(ctx).uid) {
            logger.log('GraphQL [#' + AuthContext.get(ctx).uid + ']: ' + JSON.stringify(req.body));
        } else {
            logger.log('GraphQL [#ANON]: ' + JSON.stringify(req.body));
        }
    }
    res.locals.ctx = ctx;

    const originalEnd = res.end;
    res.end = function (...args: any[]) {
        res.end = originalEnd;
        const returned = res.end.call(this, ...args);
        let span = TracingContext.get(ctx).span;
        if (span) {
            span.finish();
        }
        return returned;
    };

    next();
}