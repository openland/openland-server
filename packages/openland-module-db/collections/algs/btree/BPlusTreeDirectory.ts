import { Context } from '@openland/context';
import { encoders, Subspace, TupleItem } from '@openland/foundationdb';
import { delay } from '@openland/foundationdb/lib/utils';
import { arraySplit, LeafNode, Node, packNode, recordAdd, unpackNode } from './impl/Node';

const SUBSPACE_ROOT = 0;
const SUBSPACE_COUNTER = 1;
const SUBSPACE_NODES = 2;

export type DumpedNode = {
    type: 'internal',
    children: { count: number, min: number, max: number, node: DumpedNode }[]
} | {
    type: 'leaf',
    records: number[]
};

function isWithin(cursor: { from?: number | null, to?: number | null }, key: number) {
    if ((cursor.to !== null && cursor.to !== undefined) && (cursor.to < key)) {
        return false;
    }
    if ((cursor.from !== null && cursor.from !== undefined) && (key < cursor.from)) {
        return false;
    }
    return true;
}

function isIntervalWithin(cursor: { from?: number | null, to?: number | null }, interval: { min: number, max: number }) {
    if (cursor.from !== null && cursor.from !== undefined) {
        if (interval.min < cursor.from) {
            return false;
        }
    }
    if (cursor.to !== null && cursor.to !== undefined) {
        if (cursor.to < interval.max) {
            return false;
        }
    }
    return true;
}

function isIntersects(cursor: { from?: number | null, to?: number | null }, interval: { min: number, max: number }) {
    let from = cursor.from || Number.MIN_SAFE_INTEGER;
    let to = cursor.to || Number.MAX_SAFE_INTEGER;
    if (from > interval.max) {
        return false;
    }
    if (interval.min > to) {
        return false;
    }
    return true;
}

export class BPlusTreeDirectory {
    readonly subspace: Subspace<TupleItem[], Buffer>;
    readonly maxBranch: number;

    constructor(subspace: Subspace, maxBranch: number) {
        this.maxBranch = maxBranch;
        this.subspace = subspace.withKeyEncoding(encoders.tuple);
    }

    add = async (ctx: Context, collection: Buffer, key: number) => {
        let root = await this.readRoot(ctx, collection);
        if (!root) {
            // Allocate root node id
            let nodeId = await this.allocateNodeId(ctx, collection);

            // Write node value
            this.writeRoot(ctx, collection, nodeId);
            this.writeNode(ctx, collection, { type: 'leaf', id: nodeId, parent: null, children: [key] });
            return;
        }

        // Search from the root
        let ex = await this.treeSearch(ctx, collection, root.id, key);

        // If record already exist: exit
        if (ex.children.find((v) => v === key)) {
            return;
        }

        // Add record to node
        let newRecords = recordAdd(ex.children, key);
        this.writeNode(ctx, collection, { ...ex, children: newRecords });

        // Update counter
        if (ex.parent !== null) {
            await this.updateCount(ctx, collection, ex.parent, ex.id, newRecords.length);
            await this.updateMinMax(ctx, collection, ex.parent, ex.id, key);
        }

        // Rebalance
        await this.rebalance(ctx, collection, ex.id);
    }

    count = async (ctx: Context, collection: Buffer, cursor: { from?: number | null, to?: number | null }) => {
        let root = await this.subspace.get(ctx, [collection, SUBSPACE_ROOT]);
        if (!root) {
            return 0;
        } else {
            return this.countNode(ctx, collection, cursor, encoders.int32LE.unpack(root));
        }
    }

    private async countNode(ctx: Context, collection: Buffer, cursor: { from?: number | null, to?: number | null }, nodeId: number): Promise<number> {
        let node = await this.readNode(ctx, collection, nodeId);
        if (node.type === 'leaf') {
            let res = 0;
            for (let n of node.children) {
                if (isWithin(cursor, n)) {
                    res++;
                }
            }
            return res;
        } else if (node.type === 'internal') {
            let res = 0;
            for (let n of node.children) {
                if (isIntervalWithin(cursor, n)) {
                    res += n.count;
                } else {
                    if (isIntersects(cursor, n)) {
                        res += await this.countNode(ctx, collection, cursor, n.node);
                    }
                }
            }
            return res;
        } else {
            throw Error('Invalid type');
        }
    }

    //
    // Debug dumping
    //

    dump = async (ctx: Context, collection: Buffer): Promise<DumpedNode | null> => {
        let root = await this.subspace.get(ctx, [collection, SUBSPACE_ROOT]);
        if (!root) {
            return null;
        } else {
            return await this.dumpNode(ctx, collection, encoders.int32LE.unpack(root));
        }
    }

    private async dumpNode(ctx: Context, collection: Buffer, nodeId: number): Promise<DumpedNode> {
        let node = await this.readNode(ctx, collection, nodeId);
        if (node.type === 'leaf') {
            return {
                type: 'leaf',
                records: node.children
            };
        } else if (node.type === 'internal') {
            return {
                type: 'internal',
                children: await Promise.all(node.children.map(async (ch) => ({ min: ch.min, max: ch.max, count: ch.count, node: await this.dumpNode(ctx, collection, ch.node) })))
            };
        } else {
            throw Error('Unable to dump node');
        }
    }

    //
    // Tree Operations
    //

    private rebalance = async (ctx: Context, collection: Buffer, nodeId: number) => {
        let node = await this.readNode(ctx, collection, nodeId);
        if (node.children.length >= this.maxBranch - 1) {
            await this.splitNode(ctx, collection, node.id);
        }
    }

    private splitNode = async (ctx: Context, collection: Buffer, nodeId: number) => {
        let node = await this.readNode(ctx, collection, nodeId);

        // Resolve new parent
        let parentId: number;
        if (node.parent === null) {
            parentId = await this.allocateNodeId(ctx, collection);
        } else {
            parentId = node.parent;
        }

        // Allocate new node
        let newNodeId = await this.allocateNodeId(ctx, collection);

        let leftCount: number;
        let rightCount: number;
        let leftMin: number;
        let leftMax: number;
        let rightMin: number;
        let rightMax: number;

        if (node.type === 'leaf') {
            let split = arraySplit(node.children);

            // Write left node
            this.writeNode(ctx, collection, {
                ...node,
                parent: parentId,
                children: split.left
            });

            // Write right node
            this.writeNode(ctx, collection, {
                type: 'leaf',
                id: newNodeId,
                parent: parentId,
                children: split.right
            });

            // Resolve left metrics
            leftCount = split.left.length;
            leftMin = split.left[0];
            leftMax = split.left[split.left.length - 1];

            // Resolve right metrics
            rightCount = split.right.length;
            rightMin = split.right[0];
            rightMax = split.right[split.right.length - 1];
        } else if (node.type === 'internal') {
            let split = arraySplit(node.children);

            // Write left node
            this.writeNode(ctx, collection, {
                ...node,
                parent: parentId,
                children: split.left
            });

            // Write right node
            this.writeNode(ctx, collection, {
                type: 'internal',
                id: newNodeId,
                parent: parentId,
                children: split.right
            });

            // Update parent for new items
            for (let l of split.right) {
                await this.updateParent(ctx, collection, l.node, newNodeId);
            }

            // Resolve left metrics
            leftCount = split.left.reduce((v, i) => v + i.count, 0) /* Sum count in left */;
            leftMin = split.left[0].min;
            leftMax = split.left[split.left.length - 1].max;

            // Resolve right metrics
            rightCount = split.right.reduce((v, i) => v + i.count, 0) /* Sum count in right */;
            rightMin = split.right[0].min;
            rightMax = split.right[split.right.length - 1].max;
        } else {
            throw Error('Invalid node');
        }

        // Write parent node
        if (node.parent === null) {
            this.writeRoot(ctx, collection, parentId);
            this.writeNode(ctx, collection, {
                type: 'internal',
                id: parentId,
                parent: null,
                children: [
                    { node: node.id, min: leftMin, max: leftMax, count: leftCount },
                    { node: newNodeId, min: rightMin, max: rightMax, count: rightCount }
                ]
            });
        } else {
            // Load parent node
            let parentNode = await this.readNode(ctx, collection, node.parent);
            if (parentNode.type !== 'internal') {
                throw Error('Broken tree');
            }
            let index = parentNode.children.findIndex((v) => v.node === node.id);
            if (index < 0) {
                throw Error('Broken tree');
            }

            // Insert new children to parent
            let childrenLeft = parentNode.children.slice(0, index);
            let childrenRight = parentNode.children.slice(index + 1);
            let children: { node: number, min: number, max: number, count: number }[] = [
                ...childrenLeft,
                { node: node.id, min: leftMin, max: leftMax, count: leftCount },
                { node: newNodeId, min: rightMin, max: rightMax, count: rightCount },
                ...childrenRight
            ];
            this.writeNode(ctx, collection, {
                type: 'internal',
                id: parentNode.id,
                parent: parentNode.parent,
                children
            });
        }

        // Rebalance parent
        await this.rebalance(ctx, collection, parentId);
    }

    private treeSearch = async (ctx: Context, collection: Buffer, nodeId: number, key: number): Promise<LeafNode> => {
        let nodeRaw = await this.subspace.get(ctx, [collection, SUBSPACE_NODES, nodeId]);
        if (!nodeRaw) {
            throw Error('Unable to find node');
        }
        let node = unpackNode(nodeRaw);
        if (node.type === 'leaf') {
            return node;
        } else if (node.type === 'internal') {
            if (key < node.children[1].min) {
                return this.treeSearch(ctx, collection, node.children[0].node, key);
            } else if (node.children[node.children.length - 1].min <= key) {
                return this.treeSearch(ctx, collection, node.children[node.children.length - 1].node, key);
            } else {
                for (let i = 1; i < node.children.length - 1; i++) {
                    if (node.children[i].min <= key && node.children[i + 1].min < key) {
                        return this.treeSearch(ctx, collection, node.children[i].node, key);
                    }
                }
            }
            throw Error('Invalid node');
        }
        throw Error('Invalid node');
    }

    //
    // Node Operations
    //

    private async allocateNodeId(ctx: Context, collection: Buffer) {
        let counterRaw = await this.subspace.get(ctx, [collection, SUBSPACE_COUNTER]);
        let nodeId = 1;
        if (counterRaw) {
            nodeId = encoders.int32LE.unpack(counterRaw) + 1;
            this.subspace.set(ctx, [collection, SUBSPACE_COUNTER], encoders.int32LE.pack(nodeId));
        } else {
            this.subspace.set(ctx, [collection, SUBSPACE_COUNTER], encoders.int32LE.pack(1));
        }
        return nodeId;
    }

    private async readNode(ctx: Context, collection: Buffer, nodeId: number) {
        let node = await this.subspace.get(ctx, [collection, SUBSPACE_NODES, nodeId]);
        if (!node) {
            throw Error('Unable to find node');
        }
        return unpackNode(node);
    }

    private writeNode(ctx: Context, collection: Buffer, node: Node) {
        this.subspace.set(ctx, [collection, SUBSPACE_NODES, node.id], packNode(node));
    }

    private writeRoot(ctx: Context, collection: Buffer, nodeId: number | null) {
        if (nodeId !== null) {
            this.subspace.set(ctx, [collection, SUBSPACE_ROOT], encoders.int32LE.pack(nodeId));
        } else {
            this.subspace.clear(ctx, [collection, SUBSPACE_ROOT]);
        }
    }

    private async updateParent(ctx: Context, collection: Buffer, nodeId: number, parentId: number) {
        let node = await this.readNode(ctx, collection, nodeId);
        node.parent = parentId;
        this.writeNode(ctx, collection, node);
    }

    private async readRoot(ctx: Context, collection: Buffer): Promise<Node | null> {
        let root = await this.subspace.get(ctx, [collection, SUBSPACE_ROOT]);
        if (!root) {
            return null;
        } else {
            return this.readNode(ctx, collection, encoders.int32LE.unpack(root));
        }
    }

    private async updateCount(ctx: Context, collection: Buffer, nodeId: number, childrenId: number, count: number) {
        let node = await this.readNode(ctx, collection, nodeId);
        if (node.type !== 'internal') {
            throw Error('Node type invalid');
        }
        let index = node.children.findIndex((v) => v.node === childrenId);
        if (index < 0) {
            await delay(100);
            throw Error('Unable to find children');
        }
        node.children[index].count = count;
        this.writeNode(ctx, collection, node);

        if (node.parent !== null) {
            let sum = 0;
            for (let ch of node.children) {
                sum += ch.count;
            }
            await this.updateCount(ctx, collection, node.parent, nodeId, sum);
        }
    }

    private async updateMinMax(ctx: Context, collection: Buffer, nodeId: number, childrenId: number, key: number) {
        let node = await this.readNode(ctx, collection, nodeId);
        if (node.type !== 'internal') {
            throw Error('Node type invalid');
        }
        let index = node.children.findIndex((v) => v.node === childrenId);
        if (index < 0) {
            throw Error('Unable to find children');
        }
        node.children[index].min = Math.min(node.children[index].min, key);
        node.children[index].max = Math.max(node.children[index].max, key);
        this.writeNode(ctx, collection, node);

        if (node.parent !== null) {
            await this.updateMinMax(ctx, collection, node.parent, nodeId, key);
        }
    }
}