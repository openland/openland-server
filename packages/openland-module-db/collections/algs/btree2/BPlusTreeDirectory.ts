import { Context } from '@openland/context';
import { Subspace } from '@openland/foundationdb';
import { TreeNode, TreeNodeType } from './storage/BTree';
import { BTreeOperations } from './storage/BTreeOperations';
import { BTreeRepository } from './storage/BTreeRepository';

export class BPlusTreeDirectory {
    readonly maxBranch: number;
    readonly minChildren: number;
    readonly maxChildren: number;
    readonly store: BTreeRepository;
    readonly ops: BTreeOperations;

    constructor(subspace: Subspace, maxBranch: number) {
        this.maxBranch = maxBranch;
        this.minChildren = Math.floor(maxBranch / 2);
        this.maxChildren = maxBranch - 1;
        this.store = new BTreeRepository(subspace);
        this.ops = new BTreeOperations(this.store);
    }

    add = async (ctx: Context, collection: Buffer, key: number) => {
        let root = await this.store.getRoot(ctx, collection);
        if (!root) {

            // Allocate root node id
            let nodeId = await this.store.allocateNodeId(ctx, collection);

            // Write root node
            await this.ops.setRootNode(ctx, collection, {
                id: nodeId,
                type: TreeNodeType.LEAF,
                values: [key]
            });
            return;
        }

        // Search from the root
        let ex = await this.ops.search(ctx, collection, key);

        // If record already exist: exit
        if (ex.values.find((v) => v === key)) {
            return;
        }

        // Add value to node
        await this.ops.addValuesToLeaf(ctx, collection, ex.id, [key]);

        // Split node if needed
        await this.splitIfNeeded(ctx, collection, ex.id);
    }

    remove = async (ctx: Context, collection: Buffer, key: number) => {
        let root = await this.store.getRoot(ctx, collection);
        if (!root) {
            return;
        }

        // Search from the root
        let ex = await this.ops.search(ctx, collection, key);

        // If record already exist: exit
        if (!ex.values.find((v) => v === key)) {
            return;
        }

        // Delete node if it is root and last value in the root
        if (ex.parent === 0 && ex.values.length === 1) {
            await this.ops.deleteNode(ctx, collection, ex.id);
            return;
        }

        // Remove value from node
        await this.ops.removeValuesFromLeaf(ctx, collection, ex.id, [key]);

        // Merge node if needed
        await this.mergeIfNeeded(ctx, collection, ex.id);
    }

    count = (ctx: Context, collection: Buffer, cursor: { from?: number | null, to?: number | null }) => {
        return this.ops.count(ctx, collection, cursor);
    }

    //
    // Rebalancing
    //

    private splitIfNeeded = async (ctx: Context, collection: Buffer, nodeId: number) => {
        let node = await this.store.readNode(ctx, collection, nodeId);
        if ((node.children.length > this.maxChildren || node.values.length > this.maxChildren)) {
            let res = await this.ops.splitNode(ctx, collection, node.id);
            await this.splitIfNeeded(ctx, collection, res.parent.id);
        }
    }

    private mergeIfNeeded = async (ctx: Context, collection: Buffer, nodeId: number) => {
        let node = await this.store.readNode(ctx, collection, nodeId);

        if (node.type === TreeNodeType.LEAF && node.values.length < this.minChildren && node.parent !== 0) {

            // Try to pick from left
            let leftId = await this.ops.findNodeSiblingLeft(ctx, collection, nodeId);
            let left: TreeNode | null = null;
            if (leftId) {
                left = await this.store.readNode(ctx, collection, leftId);
                if (left.values.length - 1 >= this.minChildren) {
                    await this.ops.moveValue(ctx, collection, left.id, nodeId, left.values[left.values.length - 1]);
                    return;
                }
            }

            // Try to pick from right
            let rightId = await this.ops.findNodeSiblingRight(ctx, collection, nodeId);
            let right: TreeNode | null = null;
            if (rightId) {
                right = await this.store.readNode(ctx, collection, rightId);
                if (right.values.length - 1 >= this.minChildren) {
                    await this.ops.moveValue(ctx, collection, right.id, nodeId, right.values[0]);
                    return;
                }
            }

            // Delete node and merge parent if needed
            await this.ops.deleteNode(ctx, collection, node.id);

            // Move values
            if (leftId) {
                await this.ops.addValuesToLeaf(ctx, collection, leftId, node.values);
            } else if (rightId) {
                await this.ops.addValuesToLeaf(ctx, collection, rightId, node.values);
            }

            // Merge parent if needed
            if (node.parent !== 0) {
                await this.mergeIfNeeded(ctx, collection, node.parent);
            }

            return;
        }

        if (node.type === TreeNodeType.INNER && node.children.length < this.minChildren && node.parent !== 0) {
            // Try to pick from left
            let leftId = await this.ops.findNodeSiblingLeft(ctx, collection, nodeId);
            let left: TreeNode | null = null;
            if (leftId) {
                left = await this.store.readNode(ctx, collection, leftId);
                if (left.children.length - 1 >= this.minChildren) {
                    await this.ops.moveNode(ctx, collection, left.children[left.children.length - 1].id, nodeId);
                    return;
                }
            }

            // Try to pick from right
            let rightId = await this.ops.findNodeSiblingRight(ctx, collection, nodeId);
            let right: TreeNode | null = null;
            if (rightId) {
                right = await this.store.readNode(ctx, collection, rightId);
                if (right.children.length - 1 >= this.minChildren) {
                    await this.ops.moveNode(ctx, collection, right.children[0].id, nodeId);
                    return;
                }
            }

            // Delete node and merge parent if needed
            await this.ops.deleteNode(ctx, collection, node.id);

            // Move values
            if (leftId) {
                await this.ops.addChildren(ctx, collection, leftId, node.children);
            } else if (rightId) {
                await this.ops.addChildren(ctx, collection, rightId, node.children);
                return;
            }

            // Merge parent if needed
            if (node.parent !== 0) {
                await this.mergeIfNeeded(ctx, collection, node.parent);
            }

            return;
        }

        // Remove inner root node
        if (node.type === TreeNodeType.INNER && node.parent === 0 && node.children.length === 1) {
            let newRoot = await this.store.readNode(ctx, collection, node.children[0].id);
            await this.ops.setRootNode(ctx, collection, { ...newRoot, parent: 0 });
        }
    }
}