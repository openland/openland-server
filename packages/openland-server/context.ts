import { CacheContext } from 'openland-module-api/CacheContext';
import { AuthContext } from 'openland-module-auth/AuthContext';
import { ContextNamespaceType, registerExtension } from '@openland/context';
import { RequestContext } from 'openland-module-api/RequestContext';

declare module '@openland/context' {
    interface Context {
        readonly auth: ContextNamespaceType<typeof AuthContext>;
        readonly req: ContextNamespaceType<typeof RequestContext>;
        readonly cache: Map<string, any>;
    }
}

registerExtension('auth', (ctx) => {
    return AuthContext.get(ctx);
});
registerExtension('req', (ctx) => {
    return RequestContext.get(ctx);
});
registerExtension('cache', (ctx) => {
    return CacheContext.get(ctx)!;
});