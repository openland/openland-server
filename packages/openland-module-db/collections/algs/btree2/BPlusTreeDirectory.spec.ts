import { createNamedContext } from '@openland/context';
import { inTx, Database } from '@openland/foundationdb';
import { BPlusTreeDirectory } from './BPlusTreeDirectory';

let root = createNamedContext('test');
const COLLECTION_0 = Buffer.from([0]);
const COLLECTION_1 = Buffer.from([1]);
const COLLECTION_2 = Buffer.from([2]);
const COLLECTION_3 = Buffer.from([3]);
const COLLECTION_4 = Buffer.from([4]);
const COLLECTION_5 = Buffer.from([5]);
const COLLECTION_6 = Buffer.from([6]);
const COLLECTION_7 = Buffer.from([7]);
const COLLECTION_8 = Buffer.from([8]);

describe('BPlusTreeDirectory', () => {

    let directory: BPlusTreeDirectory;
    beforeAll(async () => {
        let db = await Database.openTest({ name: 'bplustree-directory', layers: [] });
        directory = new BPlusTreeDirectory(db.allKeys, 5);
    });

    it('should create root node', async () => {
        await inTx(root, async (ctx) => {
            await directory.add(ctx, COLLECTION_0, 1);
        });

        let dump = await inTx(root, async (ctx) => {
            return await directory.ops.dumpAll(ctx, COLLECTION_0);
        });
        expect(dump).toMatchSnapshot();
    });

    it('should create delete root node', async () => {
        await inTx(root, async (ctx) => {
            await directory.add(ctx, COLLECTION_6, 1);
        });
        await inTx(root, async (ctx) => {
            await directory.remove(ctx, COLLECTION_6, 1);
        });
        let dump = await inTx(root, async (ctx) => {
            return await directory.ops.dumpAll(ctx, COLLECTION_6);
        });
        expect(dump).toBeNull();
    });

    it('should expand root node', async () => {
        await inTx(root, async (ctx) => {
            await directory.add(ctx, COLLECTION_1, 2);
            await directory.add(ctx, COLLECTION_1, 10);
            await directory.add(ctx, COLLECTION_1, 1);
        });

        let dump = await inTx(root, async (ctx) => {
            return await directory.ops.dumpAll(ctx, COLLECTION_1);
        });
        expect(dump).toMatchSnapshot();
    });

    it('should shrink root node', async () => {
        await inTx(root, async (ctx) => {
            await directory.add(ctx, COLLECTION_7, 2);
            await directory.add(ctx, COLLECTION_7, 10);
            await directory.add(ctx, COLLECTION_7, 1);
            await directory.remove(ctx, COLLECTION_7, 1);
        });
        let dump = await inTx(root, async (ctx) => {
            return await directory.ops.dumpAll(ctx, COLLECTION_7);
        });
        expect(dump).toMatchSnapshot();

        await inTx(root, async (ctx) => {
            await directory.remove(ctx, COLLECTION_7, 10);
        });
        dump = await inTx(root, async (ctx) => {
            return await directory.ops.dumpAll(ctx, COLLECTION_7);
        });
        expect(dump).toMatchSnapshot();

        await inTx(root, async (ctx) => {
            await directory.remove(ctx, COLLECTION_7, 2);
        });
        dump = await inTx(root, async (ctx) => {
            return await directory.ops.dumpAll(ctx, COLLECTION_7);
        });
        expect(dump).toMatchSnapshot();
    });

    it('should split root node', async () => {
        await inTx(root, async (ctx) => {
            await directory.add(ctx, COLLECTION_2, 1);
            await directory.add(ctx, COLLECTION_2, 10);
            await directory.add(ctx, COLLECTION_2, 5);
            await directory.add(ctx, COLLECTION_2, 11);
            await directory.add(ctx, COLLECTION_2, 12);
        });

        let dump = await inTx(root, async (ctx) => {
            return await directory.ops.dumpAll(ctx, COLLECTION_2);
        });
        expect(dump).toMatchSnapshot();
    });

    it('should shrink splitted root node', async () => {
        await inTx(root, async (ctx) => {
            await directory.add(ctx, COLLECTION_8, 1);
            await directory.add(ctx, COLLECTION_8, 10);
            await directory.add(ctx, COLLECTION_8, 5);
            await directory.add(ctx, COLLECTION_8, 11);
            await directory.add(ctx, COLLECTION_8, 12);
        });

        let dump = await inTx(root, async (ctx) => {
            await directory.remove(ctx, COLLECTION_8, 1);
            return await directory.ops.dumpAll(ctx, COLLECTION_8);
        });
        expect(dump).toMatchSnapshot();

        dump = await inTx(root, async (ctx) => {
            await directory.remove(ctx, COLLECTION_8, 10);
            return await directory.ops.dumpAll(ctx, COLLECTION_8);
        });
        expect(dump).toMatchSnapshot();

        dump = await inTx(root, async (ctx) => {
            await directory.remove(ctx, COLLECTION_8, 5);
            return await directory.ops.dumpAll(ctx, COLLECTION_8);
        });
        expect(dump).toMatchSnapshot();
        
    });

    it('should split root internal node', async () => {
        await inTx(root, async (ctx) => {
            await directory.add(ctx, COLLECTION_3, 1);
            await directory.add(ctx, COLLECTION_3, 10);
            await directory.add(ctx, COLLECTION_3, 5);
            await directory.add(ctx, COLLECTION_3, 11);
        });

        await inTx(root, async (ctx) => {
            await directory.add(ctx, COLLECTION_3, 12);
            await directory.add(ctx, COLLECTION_3, 145);
            await directory.add(ctx, COLLECTION_3, 113);
            await directory.add(ctx, COLLECTION_3, -1);
        });

        let dump = await inTx(root, async (ctx) => {
            return await directory.ops.dumpAll(ctx, COLLECTION_3);
        });
        expect(dump).toMatchSnapshot();
    });

    it('should split internal node', async () => {
        await inTx(root, async (ctx) => {
            await directory.add(ctx, COLLECTION_4, 1);
            await directory.add(ctx, COLLECTION_4, 10);
            await directory.add(ctx, COLLECTION_4, 5);
            await directory.add(ctx, COLLECTION_4, 11);
        });

        await inTx(root, async (ctx) => {
            await directory.add(ctx, COLLECTION_4, 12);
            await directory.add(ctx, COLLECTION_4, 145);
            await directory.add(ctx, COLLECTION_4, 113);
            await directory.add(ctx, COLLECTION_4, -1);
            await directory.add(ctx, COLLECTION_4, 114);
            await directory.add(ctx, COLLECTION_4, 115);
            await directory.add(ctx, COLLECTION_4, 118);
            await directory.add(ctx, COLLECTION_4, 119);
        });

        let dump = await inTx(root, async (ctx) => {
            return await directory.ops.dumpAll(ctx, COLLECTION_4);
        });
        expect(dump).toMatchSnapshot();
    });

    it('should count', async () => {
        await inTx(root, async (ctx) => {
            for (let i = 0; i < 100; i++) {
                await directory.add(ctx, COLLECTION_5, i);
            }
        });

        let count = await inTx(root, async (ctx) => {
            let res = await directory.count(ctx, COLLECTION_5, { from: 43, to: 60 });
            return { res };
        });
        expect(count.res).toBe(18);
    });
});