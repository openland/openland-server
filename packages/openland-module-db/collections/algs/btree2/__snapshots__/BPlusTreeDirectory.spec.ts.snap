// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BPlusTreeDirectory should create root node 1`] = `
Object {
  "id": 1,
  "parent": null,
  "type": "leaf",
  "values": Array [
    1,
  ],
}
`;

exports[`BPlusTreeDirectory should expand root node 1`] = `
Object {
  "id": 1,
  "parent": null,
  "type": "leaf",
  "values": Array [
    1,
    2,
    10,
  ],
}
`;

exports[`BPlusTreeDirectory should shrink root node 1`] = `
Object {
  "id": 1,
  "parent": null,
  "type": "leaf",
  "values": Array [
    2,
    10,
  ],
}
`;

exports[`BPlusTreeDirectory should shrink root node 2`] = `
Object {
  "id": 1,
  "parent": null,
  "type": "leaf",
  "values": Array [
    2,
  ],
}
`;

exports[`BPlusTreeDirectory should shrink root node 3`] = `null`;

exports[`BPlusTreeDirectory should shrink splitted root node 1`] = `
Object {
  "children": Array [
    Object {
      "count": 2,
      "max": 10,
      "min": 5,
      "node": Object {
        "id": 1,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          5,
          10,
        ],
      },
    },
    Object {
      "count": 2,
      "max": 12,
      "min": 11,
      "node": Object {
        "id": 3,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          11,
          12,
        ],
      },
    },
  ],
  "id": 2,
  "parent": null,
  "type": "internal",
}
`;

exports[`BPlusTreeDirectory should shrink splitted root node 2`] = `
Object {
  "id": 3,
  "parent": null,
  "type": "leaf",
  "values": Array [
    5,
    11,
    12,
  ],
}
`;

exports[`BPlusTreeDirectory should shrink splitted root node 3`] = `
Object {
  "id": 3,
  "parent": null,
  "type": "leaf",
  "values": Array [
    11,
    12,
  ],
}
`;

exports[`BPlusTreeDirectory should split internal node 1`] = `
Object {
  "children": Array [
    Object {
      "count": 5,
      "max": 11,
      "min": -1,
      "node": Object {
        "children": Array [
          Object {
            "count": 3,
            "max": 5,
            "min": -1,
            "node": Object {
              "id": 1,
              "parent": 2,
              "type": "leaf",
              "values": Array [
                -1,
                1,
                5,
              ],
            },
          },
          Object {
            "count": 2,
            "max": 11,
            "min": 10,
            "node": Object {
              "id": 3,
              "parent": 2,
              "type": "leaf",
              "values": Array [
                10,
                11,
              ],
            },
          },
        ],
        "id": 2,
        "parent": 7,
        "type": "internal",
      },
    },
    Object {
      "count": 7,
      "max": 145,
      "min": 12,
      "node": Object {
        "children": Array [
          Object {
            "count": 2,
            "max": 113,
            "min": 12,
            "node": Object {
              "id": 4,
              "parent": 8,
              "type": "leaf",
              "values": Array [
                12,
                113,
              ],
            },
          },
          Object {
            "count": 2,
            "max": 115,
            "min": 114,
            "node": Object {
              "id": 5,
              "parent": 8,
              "type": "leaf",
              "values": Array [
                114,
                115,
              ],
            },
          },
          Object {
            "count": 3,
            "max": 145,
            "min": 118,
            "node": Object {
              "id": 6,
              "parent": 8,
              "type": "leaf",
              "values": Array [
                118,
                119,
                145,
              ],
            },
          },
        ],
        "id": 8,
        "parent": 7,
        "type": "internal",
      },
    },
  ],
  "id": 7,
  "parent": null,
  "type": "internal",
}
`;

exports[`BPlusTreeDirectory should split root internal node 1`] = `
Object {
  "children": Array [
    Object {
      "count": 3,
      "max": 5,
      "min": -1,
      "node": Object {
        "id": 1,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          -1,
          1,
          5,
        ],
      },
    },
    Object {
      "count": 2,
      "max": 11,
      "min": 10,
      "node": Object {
        "id": 3,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          10,
          11,
        ],
      },
    },
    Object {
      "count": 3,
      "max": 145,
      "min": 12,
      "node": Object {
        "id": 4,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          12,
          113,
          145,
        ],
      },
    },
  ],
  "id": 2,
  "parent": null,
  "type": "internal",
}
`;

exports[`BPlusTreeDirectory should split root node 1`] = `
Object {
  "children": Array [
    Object {
      "count": 2,
      "max": 5,
      "min": 1,
      "node": Object {
        "id": 1,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          1,
          5,
        ],
      },
    },
    Object {
      "count": 3,
      "max": 12,
      "min": 10,
      "node": Object {
        "id": 3,
        "parent": 2,
        "type": "leaf",
        "values": Array [
          10,
          11,
          12,
        ],
      },
    },
  ],
  "id": 2,
  "parent": null,
  "type": "internal",
}
`;
